Private Events

vidhishweta01 opened the pull request:
In This project authenticated user can created events,participate in any event, Every event page has the the list of attendees, Each event has a scope past to which event is sent to past event, project is built with Rails.

Full project description - [Click here](https://www.theodinproject.com/paths/full-stack-ruby-on-rails/courses/ruby-on-rails/lessons/associations)

vidhishweta01 committed:
added schema controllers and model

vidhishweta01 committed:
added views for user, events created routes

vidhishweta01 committed:
added linters

vidhishweta01 committed:
added readme, added test case, added screenshots to readme

vidhishweta01 committed:
added linters

vidhishweta01 committed:
fix linter errors

vidhishweta01 committed:
fix linter error

vidhishweta01 committed:
fix linter error

vidhishweta01 committed:
fix linter error

vidhishweta01 committed:
fix linter error

vidhishweta01 committed:
fix linter error

uryelah commented on app/controllers/events_controller.rb:
```diff
@@ -0,0 +1,65 @@
+  before_action :authenticate_user!, except: %i[index show]
+
+  def index
+    @events = Event.all
```
- [ ] The events index page should have both a list of past and future events, that means you should pass at least two instance variables from the Events#index, @future_events and @past_events.
  - Don't make the logic to query the Events model at the view, that logic should be made in the Event model with scopes and those scopes called at the Events controller and assigned to the variables mentioned above.
  Read more on Rails scopes here: https://www.rubyguides.com/2019/10/scopes-in-ruby-on-rails/

uryelah commented on app/views/events/index.html.erb:
```diff
@@ -0,0 +1,39 @@
+</div>
+
+<div class="row mb-5">
+  <% Event.future.each do |event| %>
+  <div class="col-md-4 mb-4">
+      <div class="card">
+        <div class="card-body">
+          <h4 class="card-title"><%= event.title.titleize %></h4>
+          <p class="small text-muted"><%= event.date_time.strftime("%B %-d, %Y") %></p>
+          <p class="card-text"><%= event.description.truncate(100) %></p>
+          <%= link_to 'Learn More', event_path(event), class: "btn text-light btn-block btn-danger" %>
+        </div>
+      </div>
+    </div>
+  <% end %>
+</div>
+
+<p class="text-light font-weight-bolder">PAST EVENTS</p>
+<div class="row">
+  <% Event.past.each do |event| %>
+  <div class="col-md-3 mb-4">
+      <div class="card">
+        <div class="card-body">
+          <h5 class="card-title"><%= event.title.titleize %></h5>
+          <p class="small text-muted"><%= event.date_time.strftime("%B %-d, %Y") %></p>
+          <p class="card-text"><%= event.description.truncate(50) %></p>
+          <%= link_to 'Learn More', event_path(event), class: "btn btn-outline-primary btn-block" %>
```
- [ ] Never use Active Record logic directly inside the views, that defies the whole architecture concept of the MVC structure as it mix application/model logic with view logic and makes rendering you code way slower since Rails will need to parse that model logic inside the erb file, get those files, and only then compile them into the final HTML pages.

- [ ] Make sure to also transfer those loops to helper methods and render a list of an _event partial. Same as having application logic in the views, having block logic, such as loops and conditionals, decreases the performance of your application.

uryelah commented on app/views/users/profile.html.erb:
```diff
@@ -0,0 +1,20 @@
+<h1 class="mb-4 text-light"> <%= @user.name %> </h1>
```
- [ ] Please rename this page so it follow the Rails naming conventions, this view should be called "show" and the action at the user controller #show, a show page is one that shows a single instance of a model while the index one shows a list, so since the profile page shows only a single user it should be a show page.
  - In the UI of the app itself it's ok to name this page "Profile", but to keep the code more organized let's follow the RoR conventions :-)

uryelah commented on app/models/event.rb:
```diff
@@ -0,0 +1,9 @@
+  has_many :invites, foreign_key: :attended_event_id
+  has_many :attendees, through: :invites
+
+  scope :past, -> { where('date_time < ?', Date.today).order(date_time: :desc) }
+  scope :future, -> { where('date_time > ?', Date.today).order(:date_time) }
```
- [ ] Use those scopes in your event index controller instead of repeating the query at the events index view.

uryelah commented on app/models/event.rb:
```diff
@@ -0,0 +1,9 @@
+class Event < ApplicationRecord
+  belongs_to :creator, class_name: 'User'
+  validates :title, :description, :date_time, :location, presence: true
```
- [ ] [OPTIONAL] Make your project more professional by adding specific validations to title, description, and location, checking for minimum length, maximum length, format, etc.

uryelah commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,9 @@
+class UsersController < ApplicationController
+  def profile
+    @user = User.find(params[:id])
```
- [ ] The User#index view requires a list of events created by the user so you need to define a `@created_events` variable with the list of events in this action so it can be used in the view.

uryelah commented on app/views/users/profile.html.erb:
```diff
@@ -0,0 +1,20 @@
+  <div class="col-md mb-4">
+    <p class="text-light"> <%= pluralize(@user.attended_events.count, 'event').upcase %> ATTENDED</p>
+    <div class="list-group mt-3">
+        <% @user.attended_events.each do |event| %>
+          <%= link_to event.title, event_path(event), class: "list-group-item list-group-item-action" %>
+        <% end %>
```
- [ ] Let's use the `@past_events` variable defined at the controller instead of getting them directly from `@user`.

uryelah replied:
- [ ] You should also add the list of upcoming events, `@future_events`, the user will attend.

uryelah commented on app/views/users/profile.html.erb:
```diff
@@ -0,0 +1,20 @@
+  <div class="col-md mb-4">
+    <p class="text-light"> <%= pluralize(@user.created_events.count, 'event').upcase %> CREATED</p>
+    <div class="list-group mt-3">
+        <% @user.created_events.each do |event| %>
+          <%= link_to event.title, event_path(event), class: "list-group-item list-group-item-action" %>
+        <% end %>
```
Let's use the @created_events variable defined at the controller instead of getting them directly from @user.

uryelah reviewed (changes requested):
Hi **Shweta Srivastava** , I'm Sarah your reviewer for this PR 👩‍🚀 

Good work so far in this project, you were very detailed oriented in the interface logic!
There are still some points to work on before moving on though but keep strong and code on because you are almost there!



# Required Changes ♻️ 


- [ ] Please remove all the logic from the views, mainly if/else conditionals and loops, I elaborate on this further in the comments below but that makes your application slower, so have that logic in helper methods or refactor it to use partials instead.
- [ ] Let's work on the professionalism of your commits, I know if can be frustrating to fix linter issues but don't just repeat the same commit message, that makes the commits completely opaque and not useful if you need to go back in history to debug a change.
  - Remember also to write your next commits following the formatting best practices such as capitalization and writing them in the imperative tense mood.


------------

Cheers and Happy coding!👏👏👏

Code review by **Sarah Uryelah Silva Chamorro**

(Feel free to contact me on slack by the handle of  **sarah** if you have any additional question)

vidhishweta01 committed:
made changes as per code review

vidhishweta01 committed:
added new line in user_helper as per the linters check

vidhishweta01 committed:
added new line in event_controller between def past and def future and at the end in user_helper as per the linters check

vidhishweta01 committed:
removed trailing spaces as per the linters check

vidhishweta01 commented:
As per the code review following things have been accomplished:
1. Removed active records from view and added @events_past and @events_future variables in event controller
2. rename the profile as show method in user controller and changes made in routes.rb accordingly
3. added methods attended_events and created_events in user_helper.rb and remove @user.created_events and @user.attended_events from the related view

vidhishweta01 committed:
remove @user.created_events from the view and added method created_event in user_helper.rb

vidhishweta01 committed:
remove @user.created_events from the view and added method created_event in user_helper.rb

mimipeshy commented:
## STATUS : CHANGES REQUESTED 🔴 ♻️ 
Hello @vidhishweta01 👋🏾 👋🏾 congratulations on making it this far 😄 . 

#### Highlights

- [ ] You have worked on the feedback given by the previous TSE ✔️ 

#### To improve
The following are yet to be implemented ⬇️ 
- [ ] In views/users/show.html.erb, you should also add the list of upcoming events, @future_events, the user will attend.

That's all 😄, kindly submit another review when done
If you need any clarification, kindly post your comment here and tag me @mimipeshy and will be glad to help

Regards,
Peris Ndanu

vidhishweta01 commented:
Upcoming Events is future events as mentioned in the scope, It is taken by @event_future 


![Screenshot from 2021-04-19 17-02-38](https://user-images.githubusercontent.com/73999611/115229921-3cc5d680-a131-11eb-92dd-e30a5b20d65d.png)
![Screenshot from 2021-04-19 17-04-10](https://user-images.githubusercontent.com/73999611/115229942-464f3e80-a131-11eb-9279-6f693346087f.png)

oloomoses reviewed (approved):
# Congratulations, Project Approved. :muscle: :tada: :clap: 
**Great job working on the suggested changes from the previous review.
You have passed all the requirements for this project :muscle: 
Feel free to merge and close this pull request.**

**I wish you the best in your next project and see you at the top :rocket:** 
#### Happy Coding
_Slack: Oloo Moses_

vidhishweta01 commented:
Thank you all of your valuable feedback I learn a lot
